merlion/models/layers.py

```diff
@@ -0,0 +1,253 @@
+        elif config is None:
+            config = cls.config_class(model=model, **kwargs)
+        return config
+
+    @property
+    def model(self):
+        return self.config.model
+
+    @model.setter
+    def model(self, model):
+        self.config.model = model
+
+    @property
+    def base_model(self):
+        return self.config.base_model
+
+    def reset(self):
+        self.model.reset()
+        self.__init__(config=self.config)
+
+    def __getstate__(self):
+        state = super().__getstate__()
+        state["model"] = self.model.__getstate__()
+        return state
+
+    def __setstate__(self, state):
+        if "model" in state:
+            model_state = state.pop("model")
+            self.model.__setstate__(model_state)
+        super().__setstate__(state)
+
+    def __reduce__(self):
+        state_dict = self.__getstate__()
+        config = state_dict.pop("config")
+        return getattr(self.__class__, "_original_cls", self.__class__), (config,), state_dict
+
+    def _save_state(self, state_dict: Dict[str, Any], filename: str = None, **save_config) -> Dict[str, Any]:
+        # don't save config for any of the sub-models
+        sub_state = state_dict
+        for d in range(self.config.depth):
+            sub_state.pop("config", None)
+            sub_state = sub_state["model"]
+        return super()._save_state(state_dict, filename, **save_config)
+
+    def __getattr__(self, item):
+        """
+        We can get callable attributes from the base model.
+        """
+        base_model = self.base_model
+        attr = getattr(base_model, item, None)
+        if callable(attr):
+            return attr
+        return self.__getattribute__(item)
+
+    def train(self, train_data: TimeSeries, *args, **kwargs):
+        train_data = self.train_pre_process(
+            train_data, require_even_sampling=self.require_even_sampling, require_univariate=self.require_univariate
+        )
+        return self.model.train(train_data, *args, **kwargs)
+
+
+class LayeredDetector(LayeredModel, DetectorBase):
+    """
+    Base class for a layered anomaly detector. Only to be used as a subclass.
+    """
+
+    def get_anomaly_score(self, time_series: TimeSeries, time_series_prev: TimeSeries = None) -> TimeSeries:
+        time_series, time_series_prev = self.transform_time_series(time_series, time_series_prev)
+        return self.model.get_anomaly_score(time_series, time_series_prev)
+
+
+class LayeredForecaster(LayeredModel, ForecasterBase):
+    """
+    Base class for a layered forecaster. Only to be used as a subclass.
+    """
+
+    def forecast(self, time_stamps, time_series_prev: TimeSeries = None, *args, **kwargs):
+        if time_series_prev is not None:
+            time_series_prev = self.transform(time_series_prev)
+        return self.model.forecast(time_stamps, time_series_prev, *args, **kwargs)
+
+
+class LayeredForecastingDetector(LayeredForecaster, ForecastingDetectorBase):
```

Comment:
Are there any plans to add implementations to this, or is this intended to be a marker class?

Reply:
The implementations are inherited from superclasses.